internal/drivers/manager.go

```diff
@@ -309,13 +310,15 @@ func (m *Manager) Provision(ctx context.Context, poolName, serverName, liteEngin
 	inst.State = types.StateInUse
 	err = m.instanceStore.Update(ctx, inst)
 	if err != nil {
+		pool.Unlock()
 		return nil, fmt.Errorf("provision: failed to tag an instance in %q pool: %w", poolName, err)
 	}
+	pool.Unlock()
 
 	// the go routine here uses the global context because this function is called
 	// from setup API call (and we can't use HTTP request context for async tasks)
 	go func(ctx context.Context) {
-		_ = m.buildPoolWithMutex(ctx, pool)
+		_, _ = m.setupInstance(ctx, pool, false)
```

Comment:
This is changing the behavior a little bit: previously, for each init call, looks like the pool would be rebuilt again and any extra instances removed. Now we only setup a single instance. Is there some other background thread which is taking care of maintaining the pool?

Reply:
instance purger takes care of it. Ideally deletion from pool is not required at the time of provisioning request

Reply:
This is creating instances as well right? Let's say there are x instances in the pool and we get x parallel requests. They will each pick up a warm instance. After that it seems creation is happening only one by one with the newer code?

Reply:
that will happen in parallel since new vm creation is via goroutines

Reply:
Tested with scale testing of 1000 parallel builds and no issues were observed.

Reply:
Nice job testing it, how long did you let the runner run for. Did you observe the behaviour of the pools over a longer period (1hr) ?

Reply:
@tphoney we did this as part of scale testing for hosted VMs. It was triggering ~100 builds/minute for around 30 minutes. Once all the builds were done, we saw that the pool count was back to the pool size.

With this PR, we're either creating a VM from the pool (if it doesn't exist). Or we use an instance from the pool and create another instance at the same time. The pool is created initially and thus the count is maintained with the completion of all the builds as well.

This works for the greedy strategy (which is currently being used) but it would not maintain the minMax strategy since currently the pool gets maintained on every call (and extra VMs removed). If we're not planning to support that (and start honoring limit in the pool yaml), this behaviour should be fine.